gflownet/utils/batch.py

```diff
@@ -0,0 +1,346 @@
+        self.states_policy = self.states2policy()
+
+    def states2policy(self, states=None, env_ids=None):
+        """
+        Converts states from a list of states in gflownet format to a tensor of states in policy format
+        states: list of gflownet states,
+        env_ids: list of env ids indicating which env corresponds to each state in states list
+
+        Returns: torch tensor of stattes in policy format
+        """
+        if states is None:
+            states = self.states
+            env_ids = self.env_ids
+        elif env_ids is None:
+            # if states are provided, env_ids should be provided too
+            raise Exception(
+                "env_ids must be provided to the batch for converting provided states to the policy format"
+            )
+        env = self._get_first_env()
+        if env.conditional:
+            states_policy = torch.zeros(
+                (states.shape[0], env.policy_input_dim),
+                device=self.device,
+                dtype=self.float,
+            )
+            for env_id in torch.unique(env_ids):
+                states_policy[env_ids == env_id] = self.envs[
+                    env_id.item()
+                ].statetorch2policy(states[env_ids == env_id])
+        else:
+            states_policy = env.statetorch2policy(states)
+        return states_policy
+
+    def states2proxy(self, states=None, env_ids=None):
+        """
+        Converts states from a list of states in gflownet format to a tensor of states in proxy format
+        states: list of gflownet states,
+        env_ids: list of env ids indicating which env corresponds to each state in states list
+
+        Returns: list of stattes in proxy format
+        """
+        if states is None:
+            states = self.states
+            env_ids = self.env_ids
+        elif env_ids is None:
+            # if states are provided, env_ids should be provided too
+            raise Exception(
+                "env_ids must be provided to the batch for converting provided states to the policy format"
+            )
+        env = self._get_first_env()
+        if env.conditional:
+            states_proxy = []
+            index = torch.arange(states.shape[0], device=self.device)
+            perm_index = []
+            for env_id in torch.unique(env_ids):
+                states_proxy.append(
+                    self.envs[env_id.item()].statetorch2proxy(states[env_ids == env_id])
+                )
+                perm_index.append(index[env_ids == env_id])
+            perm_index = torch.cat(perm_index)
+            # reverse permutation to make it index the states_proxy array
+            index[perm_index] = index.clone()
+            states_proxy = concat_items(states_proxy, index)
+        else:
+            states_proxy = env.statetorch2proxy(states)
+        return states_proxy
+
+    def _process_parents(self):
```

Comment:
Similar as my comment [above](https://github.com/alexhernandezgarcia/gflownet/pull/111/files#r1240217145) - I would return the relevant variables.

Reply:
I'm not sure how to implement it here because this method changes different attributes depending on the loss (`parents_all_policy` and `parents_all` in the flowmatch case and `parents` and `parents_policy` in the trajectory_balance case)

Reply:
I would return the four variables and make the unused ones empty lists. What do you think?

Reply:
well, I'm not strongly against, but it looks heavy to me, because then with new losses we'll potentially need to return even more things...

Reply:
I don't think the list of returned things will grow (much). But if we do not return stuff here, then we should not return it anywhere. I leave it up to you :) I also do not feel strongly about either way.

Reply:
[Done](https://github.com/alexhernandezgarcia/gflownet/pull/111/commits/7eb0f53a03ef1a5734cff7a8bb77a67e79f2776a)